### PR TITLE
ffmpeg 5.0

### PR DESCRIPTION
**Pull request by Tyrubias**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by carlocab** [comment]
The linkage failures will need to be revision bumped. Only the linkage failures for now, please. (i.e. leave the `brew test` failures alone first.)

Try
```
brew bump-revision --message "(ffmpeg 5.0)" <names of formulae to bump>
```

**Comment by carlocab** [comment]
Oh, when you do that please rebase (`git rebase`) to remove the merge commit; @BrewTestBot is unable to merge PRs that have merge commits.

**Comment by Tyrubias** [comment]
@carlocab To confirm, should I only bump the formulae that failed the linkage test but not the ones that failed both the linkage test and the `brew test`?

**Comment by robd003** [comment]
@nandahkrishna is there any way to get more verbose output for why some of the tests failed?

**Comment by carlocab** [comment]
You can expand the log in places like this one: https://github.com/Homebrew/homebrew-core/runs/4830304684?check_suite_focus=true#step:6:49

**Comment by Tyrubias** [comment]
Some of these failed because upstream doesn't support `ffmpeg` v5 yet. Should we create a versioned `ffmpeg@5` formula for this instead @carlocab?

**Comment by SMillerDev** [comment]
Definitely not a versioned `@5`. But you can create issues upstream, if they don't have a quick fix for it we can add a versioned `@4` for those formulae.

**Comment by Tyrubias** [comment]
@SMillerDev I've taken a look at about a third of these and it seems to be a nontrivial fix. FFmpeg 5 removed quite a few deprecated functions that these formula depend on. I'll still open issues but I think we should create a versioned formula. What are the steps for doing so?

**Comment by SMillerDev** [comment]
Just copy the old formula to a `ffmpeg@4.rb` file, remove the bottle block, rename the class and you're good to go.

**Comment by carlocab** [comment]
Yes, we can add an `ffmpeg@4` formula and update `ffmpeg` to 5.0.

**Comment by Tyrubias** [comment]
Should I update the broken formula to depend on `ffmpeg@4`?

**Comment by carlocab** [comment]
> Should I update the broken formula to depend on `ffmpeg@4`?

Yes, that's the idea. If there are still remaining linkage failures (which will often also result in a `test` failure), let's try just a `revision` bump first before switching them to `ffmpeg@4`.

Try to keep this PR to a single commit per formula (and a single formula per commit) so we don't have to struggle with merging this later on.

**Comment by Tyrubias** [comment]
I've run into an issue where `mlt` and `synfig` have "conflicting version recursive dependencies". `mlt` doesn't need `ffmpeg@4` but it depends on `opencv` which does. I'm planning on changing it to `ffmpeg@4` even though it doesn't strictly need the older version to fix this issue. This should also fix the `synfig` issue, as it implicitly uses the `ffmpeg` version that `mlt` uses. Is this acceptable?

**Comment by Tyrubias** [comment]
Another issue I'm running into is that `brew audit --strict ffmpeg@4` complains that `Non-executables were installed to "/opt/homebrew/opt/ffmpeg@4/bin".`. However, the versioned formula still contains the code necessary from before to fix this issue. I'm not sure if Homebrew handles installing binaries for keg-only formula differently. Could someone enlighten me on how to fix this issue?

File: Formula/ffmpeg@4.rb
Comment (carlocab) [inline Formula/ffmpeg@4.rb:9]:
```suggestion
```

File: Formula/ffmpeg@4.rb
Comment (carlocab) [inline Formula/ffmpeg@4.rb:13]:
```suggestion
    regex(/href=.*?ffmpeg[._-]v?(4(?:\.\d+)+)\.t/i)
```

File: Formula/ffmpeg@4.rb
Comment (carlocab) [inline Formula/ffmpeg@4.rb:74]:
```suggestion
      --host-ldflags=#{ENV.ldflags}
      --enable-avresample
      --enable-ffplay
```

File: Formula/ffmpeg@4.rb
Comment (carlocab) [inline Formula/ffmpeg@4.rb:117]:
```suggestion
```

**Review by carlocab — COMMENTED** [review]
Nice work, @Tyrubias. Just some suggestions re `ffmpeg@4`. Please amend the commit message for `ffmpeg@4` to read `ffmpeg@4 4.4.1 (new formula)`.

All the formulae that switched dependency to `ffmpeg@4` also need their revision bumped.

Please keep this PR to still having only one-commit-per-formula (and one-formula-per-commit). Or, if you need help with this, just make sure each of your commits modifies only one formula each, even if there are multiple commits modifying a single formula and I'll help you clean it up. You'll need to do `git pull --force` though if you need to make further changes after I clean up.

**Comment by carlocab** [comment]
> Another issue I'm running into is that `brew audit --strict ffmpeg@4` complains that `Non-executables were installed to "/opt/homebrew/opt/ffmpeg@4/bin".`. However, the versioned formula still contains the code necessary from before to fix this issue. I'm not sure if Homebrew handles installing binaries for keg-only formula differently. Could someone enlighten me on how to fix this issue?

What are the non-executables being installed in `bin` that `brew audit` is complaining about?

**Comment by Tyrubias** [comment]
> > Another issue I'm running into is that `brew audit --strict ffmpeg@4` complains that `Non-executables were installed to "/opt/homebrew/opt/ffmpeg@4/bin".`. However, the versioned formula still contains the code necessary from before to fix this issue. I'm not sure if Homebrew handles installing binaries for keg-only formula differently. Could someone enlighten me on how to fix this issue?
> 
> What are the non-executables being installed in `bin` that `brew audit` is complaining about?

`brew audit` is complaining about `/opt/homebrew/opt/ffmpeg@4/bin/python`, which is a folder with some Python example code. I think the line `mv bin/"python", pkgshare/"python", force: true` is supposed to fix it, but it still complains on my local machine.

**Comment by Tyrubias** [comment]
I think this PR might need extended CI time — it's changing a lot of formulae 😅

**Comment by unitof** [comment]
@Tyrubias you're a trooper.

**Comment by Tyrubias** [comment]
It appears the only issues left are the audit issue with `ffmpeg@4` and the failure of the `moc` formula to build. I think I might be able to change the `ffmpeg@4` formula to work. However, the `moc` formula appears to be unable to download the source code from the original website. Is there a procedure we should follow to deal with this?

**Comment by carlocab** [comment]
> **EDIT**: `moc`'s homepage warns that the official FTP link they provide could be slow. Is there any way to extend `curl` timeouts when doing `brew install -s`?

Unfortunately not. We could try adding a `mirror` if there is a suitable one, but the `url` audit might still fail.

**Comment by Tyrubias** [comment]
I've been looking around but I don't think there's a reliable mirror. Are we blocked on this issue? I'm not sure what the proper procedure is.

**Comment by Tyrubias** [comment]
I'm trying to debug the root cause of the audit issue and I think I know the issue — it appears the `pkshare` directory isn't created for versioned formula. Is this intentional?

**Comment by carlocab** [comment]
> I've been looking around but I don't think there's a reliable mirror. Are we blocked on this issue? I'm not sure what the proper procedure is.

Not necessarily. I suggest dropping `moc` from this PR for now. It'll break, and that's unfortunate, but I'm not sure it's worth delaying an upgrade to the ~150,000 `ffmpeg` monthly installs to save the ~80 `moc` monthly installs.

CC @Homebrew/core though for better ideas here.

> I'm trying to debug the root cause of the audit issue and I think I know the issue — it appears the `pkshare` directory isn't created for versioned formula (e.g. `/opt/homebrew/Cellar/ffmpeg@4/4.4.1/ffmpeg@4` does not exist but `/opt/homebrew/Cellar/ffmpeg@4/4.4.1/ffmpeg` does. Is this intentional?

I'm not sure. That is strange. I think replacing the `mv` call with
```ruby
pkgshare.install bin/"python"
```
will fix your problem, though.

**Comment by MikeMcQuaid** [comment]
> Not necessarily. I suggest dropping `moc` from this PR for now. It'll break, and that's unfortunate, but I'm not sure it's worth delaying an upgrade to the ~150,000 `ffmpeg` monthly installs to save the ~80 `moc` monthly installs.

agreed 👍🏻

**Comment by Tyrubias** [comment]
Alright, I removed the `moc` commit from this pull request. Hopefully the checks pass now.

**Comment by Tyrubias** [comment]
I'm a little confused — the dependents test for `ffmpeg@4` is still failing some linkage tests. Is there anyway I can debug this?

**Comment by carlocab** [comment]
The linkage failures are `moc` (expected), `synfig`, and `fceux`. Both `synfig` and `fceux` still need revision bumps (and potentially switching to `ffmpeg@4` if that doesn't work):
```
brew bump-revision --message="(ffmpeg 5.0)" synfig fceux
```

**Comment by Tyrubias** [comment]
> The linkage failures are `moc` (expected), `synfig`, and `fceux`. Both `synfig` and `fceux` still need revision bumps (and potentially switching to `ffmpeg@4` if that doesn't work):
> 
> ```
> brew bump-revision --message="(ffmpeg 5.0)" synfig fceux
> ```

Oops I did not realize that those still had to be bumped — they built fine from source on my computer and I forgot the bottles would have linkage issues.

Would anyone have an `x86_64` machine and/or bandwidth to test why `chromaprint` is failing to link when building from source? I only have an Apple Silicon laptop and it seems to build fine with `ffmpeg@4` on it...

**Comment by carlocab** [comment]
For `chromaprint`:
```
/usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++ -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  -sectcreate __TEXT __info_plist /tmp/chromaprint-20220120-59508-1lyjhrg/chromaprint-1.5.1/build/src/cmd/fpcalc.Info.plist CMakeFiles/fpcalc.dir/fpcalc.cpp.o -o fpcalc  -Wl,-rpath,/tmp/chromaprint-20220120-59508-1lyjhrg/chromaprint-1.5.1/build/src -lm ../libchromaprint.1.5.1.dylib -lm -framework Accelerate /usr/local/lib/libavformat.dylib /usr/local/lib/libavcodec.dylib /usr/local/lib/libavutil.dylib /usr/local/lib/libswresample.dylib -lz 
Undefined symbols for architecture x86_64:
  "_avcodec_decode_audio4", referenced from:
      chromaprint::FFmpegAudioReader::Read(short const**, unsigned long*) in fpcalc.cpp.o
ld: symbol(s) not found for architecture x86_64
```
The linker is looking for the symbol `_avcodec_decode_audio4`, which, from the `4`, I'll assume is provided by `ffmpeg@4`.

However, when we look at the libraries that `clang++` is trying to link (see first line):
```
/usr/local/lib/libavformat.dylib /usr/local/lib/libavcodec.dylib /usr/local/lib/libavutil.dylib /usr/local/lib/libswresample.dylib
```
which are going to be from `ffmpeg` and not `ffmpeg@4`.

I don't know of a good way to fix this, but a hacky way is to add
```ruby
conflicts_with "ffmpeg"
```
to `chromaprint` for now, so that we get `brew unlink ffmpeg` before the `chromaprint` build, which will hopefully fix the linking issue here. We can remove the `conflicts_with` after this PR is merged.

There is a risk that this will break other builds, however. We can try to deal with that if it comes up.

**Comment by Tyrubias** [comment]
Should I add `conflicts_with "chromaprint"` to `ffmpeg` as well? I think the audit requires that they both have this stanza.

**Comment by Tyrubias** [comment]
Also, `moc` linkage issues fail CI on `11-arm64` but not on `12-arm64`. I'm unclear on why this occurs.

**Comment by carlocab** [comment]
Looks like `ffmpeg` and `ffmpeg@4` no longer link with `frei0r` and `xvid`, despite us passing flags to enable them.

@Tyrubias, do you still have the `configure` logs from when you last built either `ffmpeg` or `ffmpeg@4`? You should be able to find them at `~/Library/Logs/Homebrew/ffmpeg[@4]`. The log for the configure script and `config.log` would be useful here.

**Comment by Tyrubias** [comment]
I just ran `brew cleanup --prune=all` a couple days ago so I don't have earlier logs — I'll reinstall `ffmpeg` and `ffmpeg@4` and upload the logs soon.

**Comment by Tyrubias** [comment]
Here's a gist of the `configure` logs.

https://gist.github.com/Tyrubias/d2c546e810a0dd9d91e637b7311065ce

**Comment by Tyrubias** [comment]
Also, looking at the failures for `qmmp` on the Intel platforms, should I bump the revision for `qmmp` too?

**Comment by Tyrubias** [comment]
@carlocab It seems the macOS-related errors are either `moc` or `qmmp`-related. I don't have a macOS 11 machine but I can try bumping the `qmmp` revision. As for the Linux build errors for `synfig` and `fceux`, I cannot replicate them on my Pop_OS! 21.10 machine and it seems the error is Perl-related — any suggestions?

**Comment by danielnachun** [comment]
To fix the issues with `synfig` and `fceux`, they need to be built with GCC 11.  This can be achieved by adding 
```
depends_on "gcc"
fails_with gcc: "5"
```
to those formulae.

I think revision bumping `qmmp` and `moc` makes sense as a first check.  It is weird that there is no linkage failure on Linux or Monterey, but hopefully the revision bump solves the issue regardless.

**Comment by iMichka** [comment]
Regarding `qmmp`, there is this PR open since some time: #87545

**Comment by carlocab** [comment]
Also blocked by a `test-bot` bug that skips building bottles for the formulae that were switched to `ffmpeg@4`. I'll try to get in a fix for this ASAP.

**Comment by carlocab** [comment]
Ok, I think the `test-bot` bug should be fixed. Let's re-run this when #92024 is finished with its CI run.

**Comment by carlocab** [comment]
`mpd` is getting a rev bump in #93979. This will need a rebase and another rev bump for `mpd`.

**Comment by Tyrubias** [comment]
> `mpd` is getting a rev bump in #93979. This will need a rebase and another rev bump for `mpd`.

Should I rebase on that branch or wait until it's merged and rebase on master?

**Comment by carlocab** [comment]
Better to wait, I think.

**Comment by carlocab** [comment]
You can go ahead and rebase. Watch out for formulae you've rev bumped here but might've been rev bumped elsewhere since -- those will disappear from your PR and will need to be bumped again.

**Comment by carlocab** [comment]
Rebased for `qmmp` in #87545. Fingers crossed...

**Comment by robd003** [comment]
Looks like siril needs a ```depends_on "ffmpeg@4"```

**Comment by carlocab** [comment]
Does it definitely not work with ffmpeg 5? Right now we have a linkage failure -- maybe it'll work if recompiled against the newer ffmpeg?

**Comment by Tyrubias** [comment]
I bumped the revision for `siril` and it built on my machine. Not sure why CI didn't catch it originally.

Also, I might be wrong but it looks like a couple of other formula also cannot build on Linux using GCC 5. Should I explicitly add `gcc` (for `gcc@11`) as a requirement?

**Comment by carlocab** [comment]
> Also, I might be wrong but it looks like a couple of other formula also cannot build on Linux using GCC 5. Should I explicitly add `gcc` (for `gcc@11`) as a requirement?

Nice catch.

Adding something like
```ruby
fails_with gcc: "5" # ffmpeg is compiled with GCC
```
should be sufficient. This will actually be all formulae in this pull request... (sorry)

**Comment by carlocab** [comment]
`freerdp` and `pianod` need `revision` bumps. Save it for after the Intel runners are done, though -- the Linux runner will be tied up at #94339.

If someone can check whether these need `ffmpeg@4` or compile against the latest `ffmpeg`, that would be very helpful.

**Comment by Tyrubias** [comment]
`pianod` explicitly needs `ffmpeg@4` and `freerdp` works find on FFmpeg v5. I'll push the changes when the Intel runners finish.

**Comment by carlocab** [comment]
CI is green everywhere but on Linux. Let's let other PRs use the long-timeout label first. The self-hosted Linux runner is also occupied at the moment.

**Review by carlocab — DISMISSED** [review]
Given how long this takes to run in CI (and our current shortage of runners), I think it's safer to merge this now and deal with `freerdp` and `pianod` separately.

Waiting longer risks version bumps to the formulae that have been rev-bumped here, and that requires rebasing and further delays.

Thoughts, @SMillerDev?

**Review by carlocab — APPROVED** [review]
Right, `--no-autosquash`.

**Comment by carlocab** [comment]
`freerdp` #94423
`pianod` #94424

Will rebase them when this is closed.

**Comment by carlocab** [comment]
Thanks for sticking with this, @Tyrubias! You did an impressive amount of work. We wouldn't be able to run this project without contributors like you.

**Comment by retokromer** [comment]
FFmpeg’s assembly code is written for Intel and the Intel binaries also run on ARM (with no noticeable performance hit).

**Pull request closed without merging**
